lms/djangoapps/completion/models.py

```diff
@@ -73,37 +125,100 @@ def submit_completion(self, user, course_key, block_key, completion):
-                block_type=block_type,
-                block_key=block_key,
-                defaults={'completion': completion},
-            )
-            if not isnew and obj.completion != completion:
-                obj.completion = completion
+
+class AggregateCompletionManager(CompletionManager):
+    """
+    Custom manager for AggregateCompletion model.
+
+    """
+
+    def submit_completion(self, user, course_key, block_key, aggregation_name, earned, possible):
+        """
+        Inserts and Updates the completion AggregateCompletion for the specified record.
+
+        Parameters:
+            * user (django.contrib.auth.models.User): The user for whom the
+              completion is being submitted.
+            * course_key (opaque_keys.edx.keys.CourseKey): The course in
+              which the submitted block is found.
+            * block_key (opaque_keys.edx.keys.UsageKey): The block that has had
+              its completion changed.
+            * aggregation_name (string): The name of the aggregated blocks.
+              This is set by the level that the aggregation
+              is occurring. Possible values are "course", "chapter", "sequential", "vertical"
+            * earned (float): The positive sum of the fractional completions of all descendant blocks up
+              to the value of possible.
+            * possible (float): The total sum of the possible completion values of all descendant
+              blocks that are visible to the user. This should be a positive integer.
+
+        Return Value:
+            (BlockCompletion, bool): A tuple comprising the created or updated
+            BlockCompletion object and a boolean value indicating whether the
+            object was newly created by this call.
+
+        Raises:
+            TypeError:
+                If the wrong type is passed for the parameters.
+
+            ValueError:
+                If the value of earned is greater than possible.
+
+            django.core.exceptions.ValidationError:
+                If earned / possible results in a number that is less than 0 or greater than 1 or any float is
+                less than zero.
+
+            RuntimeError:
+                If waffle.ENABLE_COMPLETION_TRACKING is not enabled.
+
+            django.db.DatabaseError:
+                If there was a problem getting, creating, or updating the
+                BlockCompletion record in the database.
+
+                This will also be a more specific error, as described here:
+                https://docs.djangoproject.com/en/1.11/ref/exceptions/#database-exceptions.
+                IntegrityError and OperationalError are relatively common
+                subclasses.
+        """
+        self.validate(user, course_key, block_key)
+        if earned > possible:
+            raise ValueError(_('The earned cannot be larger than the possible value.'))
+        percent = earned / possible
+
+        obj, is_new = self.get_or_create(
```

Comment:
see note above about `update_or_create` and `pre_save`, which gets rid of lots of this.